File: gams/apifiles/Python/xp_example2.py
from gdxcc import *
from gamsxcc import *
from optcc import *
import sys
import os


def terminate(gdxHandle, gamsxHandle, optHandle, rc):
    optFree(optHandle)
    gdxFree(gdxHandle)
    gamsxFree(gamsxHandle)
    os._exit(rc)

def reportGDXError(gdxhandle, s):
    ret = gdxErrorStr(gdxHandle, gdxGetLastError(gdxHandle))
    print "*** Error "+ s +": " + ret[1]

def writeModelData(gdxHandle, fngdxfile):
    ret, errNr = gdxOpenWrite(gdxHandle, fngdxfile, "xp_Example1")
    if errNr:
        print "*** Error gdxOpenWrite: " + gdxErrorStr(gdxHandle, errNr)[1]
        return False
    
    if not gdxDataWriteStrStart(gdxHandle, "Demand", "Demand data", 1, GMS_DT_PAR, 0):
        reportGDXError(gdxHandle, "DataWriteStrStart")
    
    values = doubleArray(GMS_VAL_MAX)
    values[GMS_VAL_LEVEL] = 324.0
    gdxDataWriteStr(gdxHandle, ["New-York"], values)
    values[GMS_VAL_LEVEL] = 299.0
    gdxDataWriteStr(gdxHandle, ["Chicago"], values)
    values[GMS_VAL_LEVEL] = 274.0
    gdxDataWriteStr(gdxHandle, ["Topeka"], values)
    
    if not gdxDataWriteDone(gdxHandle):
        reportGDXError(gdxHandle, "WriteData")
    
    errNr = gdxGetLastError(gdxHandle)
    if errNr:
        print "*** Error while writing GDX File: " + gdxErrorStr(gdxHandle, errNr)[1]
        return False
    
    errNr = gdxClose(gdxHandle)
    if errNr:
        print "*** Error gdxClose: " + gdxErrorStr(gdxHandle, errNr)[1]
        return False
    
    return True


def callGams(gamsxHandle, optHandle, sysDir):
    deffile = sysDir + "/optgams.def"
    
    if optReadDefinition(optHandle, deffile):
        print "*** Error ReadDefinition, cannot read def file:" + deffile
        return False
    
    optSetStrStr(optHandle, "SysDir",    sysDir)
    optSetStrStr(optHandle, "Input",     "../GAMS/model2.gms")
    optSetIntStr(optHandle, "LogOption", 3)    
    ret = gamsxRunExecDLL(gamsxHandle, optHandleToPtr(optHandle), sysDir, 1)
    if ret[0] != 0:
        print "*** Error RunExecDLL: Error in GAMS call = " + str(ret[1])
        return False
    
    return True

def readSolutionData(gdxHandle, fngdxfile):
    errNr = gdxOpenRead(gdxHandle, fngdxfile)[1]
    if errNr:
        print "**** Error OpenRead: " + gdxErrorStr(gdxHandle, errNr)[1]
        return False
    
    ret, varNr = gdxFindSymbol(gdxHandle, "result")
    if not ret:
        print "*** Error FindSymbol: Could not find variable result"
        return False
    
    ret, symName, dim, varType = gdxSymbolInfo(gdxHandle, varNr)
    if dim != 2 or varType != GMS_DT_VAR:
        print "**** Error SymbolInfo: result is not a a two dimensional variable"
        return False
    
    ret, nrRecs =  gdxDataReadStrStart(gdxHandle, varNr)
    if not ret:
        reportGDXError(gdxHandle, "DataReadStrStart")
        return False
    
    for i in range(nrRecs):
        ret, elements, values, afdim = gdxDataReadStr(gdxHandle)
        if not ret:
            print "**** Error gdxDataReadStr: " + gdxErrorStr(gdxHandle, ret)[1]
            return False
        if 0 == values[GMS_VAL_LEVEL]: continue
        for d in range(dim):
            print elements[d],
            if d < dim-1:
                print ".",
        print " = " + str(values[GMS_VAL_LEVEL])
    print "All solution values shown"
    gdxDataReadDone(gdxHandle)
    gdxClose(gdxHandle)

    return True


if __name__ == "__main__":
    
    numberParams = len(sys.argv)
    if numberParams != 2 :
        print "Usage:", sys.argv[0], "sysDir"
        os._exit(1)
    
    gdxHandle   = new_gdxHandle_tp()
    optHandle   = new_optHandle_tp()
    gamsxHandle = new_gamsxHandle_tp()
    
    ok = True
    sysDir = sys.argv[1]
    print sys.argv[0], "using GAMS system directory:", sys.argv[1]

    rc = gamsxCreateD(gamsxHandle, sysDir, GMS_SSSIZE)
    assert rc[0], rc[1]
    rc = gdxCreateD  (gdxHandle,   sysDir, GMS_SSSIZE)
    assert rc[0], rc[1]
    rc = optCreateD  (optHandle,   sysDir, GMS_SSSIZE)
    assert rc[0], rc[1]

    status = writeModelData(gdxHandle, "demanddata.gdx")
    if not status:
        print("Model data not written")
        terminate(gdxHandle, gamsxHandle, optHandle, 1)
    
    status = callGams(gamsxHandle, optHandle, sysDir)
    if not status:
        print("Call to GAMS failed")
        terminate(gdxHandle, gamsxHandle, optHandle, 1)

    status = readSolutionData(gdxHandle, "results.gdx")
    if not status:
        print("Could not read solution back")
    
    terminate(gdxHandle, gamsxHandle, optHandle, 0)

File: gams/apifiles/C/xp_example2.c
/*
  Use this command to compile the example:
  cl xp_example2.c api/gdxcc.c api/optcc.c api/gamsxcc.c -Iapi
*/

/*
 This program performs the following steps:
    1. Generate a gdx file with demand data
    2. Calls GAMS to solve a simple transportation model
       (The GAMS model writes the solution to a gdx file)
    3. The solution is read from the gdx file
*/

#include "gclgms.h"
#include "gamsxcc.h"
#include "gdxcc.h"
#include "optcc.h"

#include <stdio.h>
#include <string.h>
#include <assert.h>

/* Handles to the GAMSX, GDX, and Option objects */
gamsxHandle_t Gptr=NULL;
gdxHandle_t   Xptr=NULL;
optHandle_t   Optr=NULL;

#define optSetStrS(o,os,ov) if (optFindStr(Optr,os,&optNr,&optRef)) \
                                    optSetValuesNr(Optr,optNr,0,0.0,ov)

#define gdxerror(i, s) { gdxErrorStr(Xptr, i, msg); \
                         printf("%s failed: %s\n",s,msg); return 1; }

int WriteModelData(const char *fngdxfile)
{
  int                status;
  char               msg[GMS_SSSIZE];
  gdxStrIndex_t      strIndex;
  gdxStrIndexPtrs_t  sp;
  gdxValues_t        v;

  assert(Xptr);
  gdxOpenWrite(Xptr, fngdxfile, "xp_Example2", &status);
  if (status)
    gdxerror(status, "gdxOpenWrite");

  if (0==gdxDataWriteStrStart(Xptr, "Demand", "Demand Data", 1, GMS_DT_PAR, 0))
    gdxerror(gdxGetLastError(Xptr), "gdxDataWriteStrStart");

  /* Initalize some GDX data structure */
  GDXSTRINDEXPTRS_INIT(strIndex,sp);
  strcpy(sp[0], "New-York"); v[GMS_VAL_LEVEL] = 324.0; gdxDataWriteStr(Xptr, (const char **)sp, v);
  strcpy(sp[0], "Chicago" ); v[GMS_VAL_LEVEL] = 299.0; gdxDataWriteStr(Xptr, (const char **)sp, v);
  strcpy(sp[0], "Topeka"  ); v[GMS_VAL_LEVEL] = 274.0; gdxDataWriteStr(Xptr, (const char **)sp, v);

  if (0==gdxDataWriteDone(Xptr))
    gdxerror(gdxGetLastError(Xptr), "gdxDataWriteDone");

  if (gdxClose(Xptr))
    gdxerror(gdxGetLastError(Xptr), "gdxClose");

  return 0;
}

int CallGams(const char *sysdir)
{
  char msg[GMS_SSSIZE], deffile[GMS_SSSIZE];
  int saveEOLOnly, optNr, optRef;

  assert(Gptr); assert(Optr);
  strncpy(deffile, sysdir, sizeof(deffile));
  strncat(deffile, "/optgams.def", sizeof(deffile));

  if (optReadDefinition(Optr,deffile)) {
    int i, itype;
    for (i=1; i<=optMessageCount(Optr); i++) {
      optGetMessage(Optr, i, msg, &itype);
      printf("%s\n", msg);
    }
    return 1;
  }

  saveEOLOnly = optEOLOnlySet(Optr,0);
  optReadFromPChar(Optr, "I=../GAMS/model2.gms lo=2");
  optEOLOnlySet(Optr,saveEOLOnly);
  optSetStrS(Optr,"sysdir", sysdir);

  if (gamsxRunExecDLL(Gptr, Optr, sysdir, 1, msg)) {
    printf ("Could not execute RunExecDLL: %s", msg);
    return 1;
  }

  return 0;
}

int ReadSolutionData(const char *fngdxfile)
{
  int                status, VarNr, NrRecs, FDim, dim, vartype;
  char               msg[GMS_SSSIZE], VarName[GMS_SSSIZE];
  gdxStrIndex_t      strIndex;
  gdxStrIndexPtrs_t  sp;
  gdxValues_t        v;

  assert(Xptr);
  gdxOpenRead(Xptr, fngdxfile, &status);
  if (status)
    gdxerror(status, "gdxOpenRead");

  strcpy(VarName, "result");
  if (0==gdxFindSymbol(Xptr, VarName, &VarNr)) {
    printf("Could not find variable >%s<\n", VarName);
    return 1;
  }

  gdxSymbolInfo(Xptr, VarNr, VarName, &dim, &vartype);
  if (2 != dim || GMS_DT_VAR != vartype) {
    printf("%s is not a two dimensional variable\n", VarName);
    return 1;
  }

  if (0==gdxDataReadStrStart(Xptr, VarNr, &NrRecs))
    gdxerror(gdxGetLastError(Xptr), "gdxDataReadStrStart");

  /* Initalize some GDX data structure */
  GDXSTRINDEXPTRS_INIT(strIndex,sp);

  while (gdxDataReadStr(Xptr, sp, v, &FDim)) {
    int i;

    if (0.0==v[GMS_VAL_LEVEL]) continue; /* skip level = 0.0 is default */
    for (i=0; i<dim; i++)
      printf("%s%s", sp[i],(i<dim-1)? ".":"");
    printf(" = %g\n",v[GMS_VAL_LEVEL]);
  }
  printf("All solution values shown\n");

  gdxDataReadDone(Xptr);

  if ((status=gdxGetLastError(Xptr)))
    gdxerror(status, "GDX");

  if (gdxClose(Xptr))
    gdxerror(gdxGetLastError(Xptr), "gdxClose");

  return 0;
}

int main (int argc, char *argv[])
{

  int        status;
  char       sysdir[GMS_SSSIZE], msg[GMS_SSSIZE];
  const char defsysdir[] = "c:\\Program Files\\GAMS23.5";


  if (argc > 1) {
    if(strlen(argv[1]) >= GMS_SSSIZE) {
      printf("*** Your system directory (argument 1) cannot exceed 255 characters\n");
      exit(1);
    }
    strncpy(sysdir, argv[1], sizeof(sysdir));
  }  
  else
    strncpy(sysdir, defsysdir, sizeof(sysdir));

  printf("Loading objects from GAMS system directory: %s\n", sysdir);

  /* Create objects */
  if (! gamsxCreateD (&Gptr, sysdir, msg, sizeof(msg))) {
    printf("Could not create gamsx object: %s\n", msg);
    return 1;
  }

  if (! gdxCreateD (&Xptr, sysdir, msg, sizeof(msg))) {
    printf("Could not create gdx object: %s\n", msg);
    return 1;
  }

  if (! optCreateD (&Optr, sysdir, msg, sizeof(msg))) {
    printf("Could not create opt object: %s\n", msg);
    return 1;
  }

  if ((status=WriteModelData("demanddata.gdx"))) {
    printf("Model data not written\n");
    goto TERMINATE;
  }

  if ((status=CallGams(sysdir))) {
    printf("Call to GAMS failed\n");
    goto TERMINATE;
  }

  if ((status=ReadSolutionData("results.gdx"))) {
    printf("Could not read solution back\n");
    goto TERMINATE;
  }

 TERMINATE:
  optFree(&Optr);
  gdxFree(&Xptr);
  gamsxFree(&Gptr);

  return status;
}
